include/revng/FunctionIsolation/IsolationFunctionKind.h

```diff
@@ -0,0 +1,188 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/DepthFirstIterator.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/ADT/StringRef.h"
+
+#include "revng/ABI/FunctionType.h"
+#include "revng/ADT/GenericGraph.h"
+#include "revng/Model/Binary.h"
+#include "revng/Model/IRHelpers.h"
+#include "revng/Pipeline/AllRegistries.h"
+#include "revng/Pipeline/Context.h"
+#include "revng/Pipeline/LLVMContainer.h"
+#include "revng/Pipeline/Location.h"
+#include "revng/Pipeline/Target.h"
+#include "revng/Pipes/FileContainer.h"
+#include "revng/Pipes/FunctionKind.h"
+#include "revng/Pipes/Kinds.h"
+#include "revng/Pipes/ModelGlobal.h"
+#include "revng/Pipes/TaggedFunctionKind.h"
+#include "revng/Support/FunctionTags.h"
+#include "revng/TupleTree/TupleTreeDiff.h"
+#include "revng/Yield/CrossRelations.h"
+
+namespace revng::kinds {
+
+class IsolatedFunctionKind : public TaggedFunctionKind {
+public:
+  IsolatedFunctionKind(llvm::StringRef Name,
+                       pipeline::Rank *Rank,
+                       const FunctionTags::Tag &Tag) :
+    TaggedFunctionKind(Name, Rank, Tag) {}
+
+  IsolatedFunctionKind(llvm::StringRef Name,
+                       TaggedFunctionKind &Parent,
+                       pipeline::Rank *Rank,
+                       const FunctionTags::Tag &Tag) :
+    TaggedFunctionKind(Name, Rank, Tag) {}
+
+  void getInvalidations(const pipeline::Context &Ctx,
+                        pipeline::TargetsList &ToRemove,
+                        const pipeline::GlobalTupleTreeDiff &Diff,
+                        const pipeline::Global &Before,
+                        const pipeline::Global &After) const override;
+};
+
+inline IsolatedFunctionKind
+  Isolated("Isolated", &ranks::Function, FunctionTags::Isolated);
+
+inline TaggedFunctionKind
+  ABIEnforced("ABIEnforced", &ranks::Function, FunctionTags::ABIEnforced);
+
+using yield::CrossRelations;
+
+// Find callers of a given `model::Function` via CrossRelations
```

Comment:
Is there a real reason for functions below to be in a public header?

Reply:
The idea was to expose those functions in a public header (but hide the inner implementation inside `detail` namespace) in order to make them available for `IsolatedFunctionKind.cpp` and for the upcoming `ABIEnforcedKind.cpp`.

Reply:
This might be true for some of them, but not all. Please review what you have here.
For those you think will be useful in other invalidations, create a new header with an appropriate name.
Also consider having a `.cpp` file for implementations.

Reply:
Sounds good: I added `InvalidationHelpers.h` under `Pipes` and moved those functions not needed to be exposed in the implementation file.